spi/core-spi/src/main/java/org/eclipse/dataspaceconnector/spi/telemetry/Telemetry.java

```diff
@@ -0,0 +1,84 @@
+package org.eclipse.dataspaceconnector.spi.telemetry;
+
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.context.propagation.TextMapGetter;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * System observability interface for tracing and metrics
+ */
+public class Telemetry {
+
+    private final OpenTelemetry openTelemetry;
+
+    public Telemetry(OpenTelemetry openTelemetry) {
+        this.openTelemetry = openTelemetry;
+    }
+
+    /**
+     * Gets the trace context from the current thread
+     *
+     * @return The trace context as a Map
+     */
+    public Map<String, String> getCurrentTraceContext() {
+        Map<String, String> traceContext = new HashMap<>();
+        openTelemetry.getPropagators().getTextMapPropagator()
+                .inject(Context.current(), traceContext, Map::put);
+        return traceContext;
+    }
+
+    /**
+     * Propagates the trace context present in the carrier to the current thread and returns a {@link Scope} which
+     * corresponds to the current scope of execution.
+     * <p>
+     * {@link Scope#close()} must be called to restore the previous parent context and avoid leaking child scopes.
+     * It is recommended to use try-with-resources to call {@link Scope#close()} automatically.
+     *
+     * <pre>{@code
+     * var prevTraceContext = telemetry.getCurrentTraceContext();
+     * try (Scope scope = setCurrentTraceContext(traceCarrier)) {
+     *   assert telemetry.getCurrentTraceContext() == traceCarrier.getTraceContext()
+     * }
+     * assert telemetry.getCurrentTraceContext() == prevTraceContext;
+     * }</pre>
+     *
+     * @param carrier The trace context carrier
+     */
+    public Scope setCurrentTraceContext(TraceCarrier carrier) {
```

Comment:
A setter (or setter-like) method with a return value seems a bit odd. In addition, the return value isn't used anyway, so we could just make this a normal setter.

Alternatively, seeing that the only usage of this method makes use of the Scope being an `AutoCloseable`, I'd suggest renaming this method and returning an `AutoCloseable`

Reply:
Actually this function is does not need to be exposed anymore, as we use the `contextPropagationMiddleware` method instead. I have made it private and renamed it to `propagateTraceContext`.

Reply:
Ah I just saw - `Scope#close()` does not throw an exception, while `AutoCloseable#close()` throws an `Exception`. So it's better to return `Scope` instead of `AutoCloseable`.